src/synthesis/clean.ts

```diff
@@ -0,0 +1,118 @@
+import { parse } from "discord-markdown-parser";
+import type { Message } from "discord.js";
+import type { SingleASTNode, ASTNode } from "simple-markdown";
+
+export function cleanMarkdown(message: Message) {
+  const ast = parse(message.content, "extended");
+  return text(ast, message);
+}
+
+const dateTimeFormat = new Intl.DateTimeFormat("ja-JP", {
+  dateStyle: "full",
+  timeStyle: "full",
+});
+
+function text(ast: ASTNode, message: Message): string {
+  if (Array.isArray(ast)) {
+    return ast.map((node) => text(node, message)).join("");
+  }
+
+  switch (ast.type) {
+    case "link":
+    case "blockQuote":
+    case "em":
+    case "strong":
+    case "underline":
+    case "strikethrough":
+    case "spoiler":
+      return text(astNodeOrEmpty(ast.content), message);
+
+    case "text":
+    case "escape":
+    case "inlineCode":
+      return stringOrEmpty(ast.content);
+
+    case "url":
+    case "autolink":
+      return "URL省略";
+
+    case "newline":
+    case "br":
+      return "\n";
+
+    case "codeBlock": {
+      const lang = stringOrEmpty(ast.lang);
+      return lang ? `${lang}のコード` : "コード";
+    }
+
+    case "user": {
+      const id = stringOrEmpty(ast.id);
+      const member = message.guild?.members.cache.get(id);
+      return member?.displayName ?? "不明なユーザー";
+    }
+    case "channel": {
+      const id = stringOrEmpty(ast.id);
+      const channel = message.guild?.channels.cache.get(id);
+      return channel?.name ?? "不明なチャンネル";
+    }
+    case "role": {
+      const id = stringOrEmpty(ast.id);
+      const role = message.guild?.roles.cache.get(id);
+      return role?.name ?? "不明なロール";
+    }
+    case "emoji": {
+      const id = stringOrEmpty(ast.id);
+      const emoji = message.guild?.emojis.cache.get(id);
+      return emoji?.toString() ?? "不明な絵文字";
```

Comment:
toStringだとidになってしまいます．
```suggestion
      return emoji?.name ?? "不明な絵文字";
```